File: lambdas/nodejs/memory-intensive/handler.ts
import type { Context } from "aws-lambda";
import { createHash } from "crypto";

// Fixed array size for consistent performance measurement across Lambda memory configs
const FIXED_ARRAY_SIZE_MB = 100;

interface BenchmarkEvent {
  // Event is currently unused but kept for API consistency
}

interface BenchmarkSuccess {
  success: true;
  workloadType: 'memory-intensive';
  sizeMB: number;
  architecture: string;
  memoryLimitMB: number;
  resultHash: string;
}

interface BenchmarkError {
  success: false;
  workloadType: 'memory-intensive';
  error: string;
}

type BenchmarkResult = BenchmarkSuccess | BenchmarkError;

/**
 * Lambda handler - Memory intensive memory workload benchmark allocates and sorts a fixed 
 * 100 MB array to measure performance scaling across different Lambda memory configurations.
 */
export async function handler(_event: BenchmarkEvent, context: Context): Promise<BenchmarkResult> {
  console.log(
    JSON.stringify({
      event: 'handler_start',
      workloadType: 'memory-intensive',
      runtime: `nodejs${process.version}`,
      architecture: process.arch,
      requestId: context.awsRequestId
    })
  );

  try {
    const result = memoryIntensiveWorkload(FIXED_ARRAY_SIZE_MB);

    console.log(
      JSON.stringify({
        event: 'handler_success',
        sizeMB: FIXED_ARRAY_SIZE_MB,
        arrayElements: Math.floor((FIXED_ARRAY_SIZE_MB * 1024 * 1024) / 8)
      })
    );

    return {
      success: true,
      workloadType: 'memory-intensive',
      sizeMB: FIXED_ARRAY_SIZE_MB,
      architecture: process.arch,
      memoryLimitMB: parseInt(context.memoryLimitInMB, 10),
      resultHash: result.slice(0, 64)
    };
  } catch (error) {
    console.error(
      JSON.stringify({
        event: 'handler_error',
        errorType: error instanceof Error ? error.constructor.name : 'Unknown',
        errorMessage: error instanceof Error ? error.message : String(error)
      })
    );

    return {
      success: false,
      workloadType: 'memory-intensive',
      error: error instanceof Error ? error.message : String(error)
    };
  }
}

/**
 * Allocates and sorts fixed 100 MB array to stress memory bandwidth
 *
 * Sort operation stresses both memory bandwidth (accessing all elements)
 * and CPU (comparison operations), providing comprehensive memory subsystem test.
 */
function memoryIntensiveWorkload(sizeMB: number): string {
  const arraySize = Math.floor((sizeMB * 1024 * 1024) / 8); // Float64 = 8 bytes

  // Use Float64Array for true 8-byte elements (memory parity with Python array.array('q') and Rust Vec<i64>)
  const randomNumberArray = new Float64Array(arraySize);
  for (let i = 0; i < arraySize; i++) {
    randomNumberArray[i] = Math.floor(Math.random() * 1_073_741_824);
  }

  randomNumberArray.sort();

  const sample = JSON.stringify(randomNumberArray.slice(0, 1000));
  const result = createHash('sha256').update(sample).digest('hex');
  return result;
}

File: lambdas/python/memory-intensive/handler.py
from __future__ import annotations

import array
import hashlib
import json
import logging
import platform
import random
from typing import Any

logger = logging.getLogger()
logger.setLevel(logging.INFO)

# Fixed array size for consistent performance measurement across Lambda memory configs
FIXED_ARRAY_SIZE_MB = 100


def lambda_handler(event: dict[str, Any] | None, context) -> dict[str, Any]:
    """Lambda handler - Memory intensive workload benchmark.

    Allocates and sorts a fixed 100 MB array to measure performance scaling
    across different Lambda memory configurations.
    """
    event = event or {}

    logger.info(json.dumps({
        "event": "handler_start",
        "workloadType": "memory-intensive",
        "runtime": f"python{platform.python_version()}",
        "architecture": platform.machine(),
        "requestId": getattr(context, "aws_request_id", "unknown")
    }))

    try:
        result_hash = _memory_sort(FIXED_ARRAY_SIZE_MB)

        logger.info(json.dumps({
            "event": "handler_success",
            "sizeMB": FIXED_ARRAY_SIZE_MB,
            "arrayElements": (FIXED_ARRAY_SIZE_MB * 1024 * 1024) // 8
        }))

        return {
            "success": True,
            "workloadType": "memory-intensive",
            "sizeMB": FIXED_ARRAY_SIZE_MB,
            "architecture": platform.machine(),
            "pythonVersion": platform.python_version(),
            "memoryLimitMB": int(getattr(context, "memory_limit_in_mb", 0) or 0),
            "resultHash": result_hash,
        }
    except Exception as e:
        logger.error(json.dumps({
            "event": "handler_error",
            "errorType": type(e).__name__,
            "errorMessage": str(e)
        }))
        return _fail(f"{type(e).__name__}: {e}")


def _memory_sort(size_mb: int) -> str:
    """Allocates and sorts fixed 100 MB array to stress memory bandwidth.

    Sort operation stresses both memory bandwidth (accessing all elements)
    and CPU (comparison operations), providing comprehensive memory subsystem test.
    """
    count = (size_mb * 1024 * 1024) // 8

    # 'q' = signed 64-bit integer (8 bytes), matching JavaScript's Float64Array
    data = array.array('q', (random.getrandbits(30) for _ in range(count)))

    sorted_data = array.array('q', sorted(data))

    sample = sorted_data[:1000]
    return hashlib.sha256(sample.tobytes()).hexdigest()


def _fail(msg: str) -> dict[str, Any]:
    return {"success": False, "workloadType": "memory-intensive", "error": msg}
